RoboJackets/robocup-software
# ROS 2 Debug Draw Interface

## Description
Debug drawer for ROS 2 nodes. This allows us to draw debug information to layers on the field by publishing to the `/viz/debug_draw` topic.

## Associated Issue
N/A

## Design
There is a one-to-one correspondence between the `DebugDraw` message and a layer. This means that when you publish a `DebugDraw` message, the previous contents of the layer will be cleared - but otherwise the most recent copy of each layer will be latched. This allows us to publish debug drawing information from nodes like planning that may not be exactly coupled to the GUI update frequency.

![image](https://user-images.githubusercontent.com/9097872/95143086-364ab100-0743-11eb-894f-a6c86e8ea651.png)
This will become more filled in as more ROS nodes require debug drawing.

## Steps to test
### Manual test
1. Run soccer: `ros2 launch rj_robocup sim.launch.py`
2. Run the following command to publish a debug layer:
```sh
ros2 topic pub /viz/debug_draw rj_drawing_msgs/msg/DebugDraw "{layer: 'test', debug_text: [{text: 'TEST'}]}"
```
3. Run the following command to clear the debug layer:
```sh
ros2 topic pub /viz/debug_draw rj_drawing_msgs/msg/DebugDraw "{layer: 'test'}"
```

Expected result: After step 2, the text `TEST` should appear at our goal. It should be toggleable by checking/unchecking the `test` layer in the Layers tab. After step 3, the text should be cleared.
